### DISCUSSION TITLE
V3 🎉

### DISCUSSION THREAD

**Discussion by simeydotme** [simeydotme/svelte-range-slider-pips#145] [General]
## Re-write in SvelteKit and Typescript. 
_(yay types for all you weird type nerds!)_

---

- test folders with examples of the project running in many environments (react, vue, jquery, esm modules, svelte 5)
- types
- insanely complicated build system and package json
- some types
- code utils split out

Hope this goes smoothly for y'all who want to upgrade because of the slightly annoying cautions in your build systems! 🤞

<hr /><em>This discussion was created from the release <a href='https://github.com/simeydotme/svelte-range-slider-pips/releases/tag/3.0.0'>V3 🎉</a>.</em>

**Comment by simeydotme** [comment]
If anyone has some comments/issues... drop them in here for dicussion!

I'm also going to need to re-do the entire Astro Docs site 🙈  ... oh god. 
What version is that even on now??

**Reply by Gurimarukin** [reply]
Hi!  
I'm afraid this update broke our sliders in our Svelte code base.

I tried with a very simple example :
```svelte
<script lang="ts">
  import { RangeSlider } from 'svelte-range-slider-pips'

  $: values = [56]
</script>

<RangeSlider bind:values />
<pre>values: {values}</pre>
```
But the slider refuses to slide:

https://github.com/user-attachments/assets/97c70747-c308-40f4-9388-ac5fc2e200c4

Previous version worked just fine

---

I was surprised I had to import `{ RangeSlider }`, but if I import it with:
```ts
import RangeSlider from 'svelte-range-slider-pips'
```
I get this error in my browser: `Uncaught SyntaxError: The requested module 'http://localhost:3000/node_modules/.vite/deps/svelte-range-slider-pips.js?v=0bcf7088' doesn't provide an export named: 'default'`

**Reply by simeydotme** [reply]
thanks @Gurimarukin 

Yes, I am currently going through hell trying to understand the package.json, sveltekit bundler, and typescript/intellisense...
they have made bundling very very difficult and complicated in sveltekit, sadly. Additional problems are that I can't tell, locally, if my NPM is using a cache or not. The test projects are all working fine. And I can't find anyway to validate a package.json to see what the imports resolve to :'( 

question; are you running sveltekit or typescript in your project?

**Reply by simeydotme** [reply]
I was responsible and set this a 3.0.0 breaking change, though, so for now you can set your package.json to `^2.3.0` and you should be OK until I figure this thing out

**Reply by simeydotme** [reply]
what's strange, though, is that it works on StackBlitz
https://stackblitz.com/edit/vitejs-vite-wevibs?file=package.json,src%2FApp.svelte&terminal=dev
but not on REPL
https://svelte.dev/repl/130c7c65330244dc9ba5fd74c3d92c00?version=4.2.19

I opened a thread in Svelte Discord if anyone is a god with Package.json and wants to help;
https://discord.com/channels/457912077277855764/1278430742581608551

**Reply by Gurimarukin** [reply]
We are using TypeScript, but we are not "running" it(?)  
App is run locally and built with Vite.  
So I was about to write that I don't understand what we are doing differently from StackBlitz.  
We don't have the `svelte.config.js` (but adding it doesn't change anything).  

But then, I commented our `immutable` compiler option out and it works:
![image](https://github.com/user-attachments/assets/89f48e3a-847e-46d9-a680-dc74c7e0c9e8)
¯\\\_(ツ)\_/¯

Thank you for your answer

**Reply by simeydotme** [reply]
Ahhh.... The immutable thing again... There was a issue with this in the past... Thank you!

**Reply by simeydotme** [reply]
[Immutable; https://github.com/simeydotme/svelte-range-slider-pips/issues/101](https://github.com/simeydotme/svelte-range-slider-pips/issues/101)

**Reply by simeydotme** [reply]
> I was surprised I had to import `{ RangeSlider }`, but if I import it with:
> 
> ```ts
> import RangeSlider from 'svelte-range-slider-pips'
> ```

Oh, i also fixed this so both work now, not sure if that's good or not

**Reply by kelvinsjk** [reply]
Thanks for the hard work!

Will be happy to help out with the documentation, so do give a holler if there's anything I can help with.

Some thoughts about documentation setup:

- Since the library is now built with SvelteKit packaging feature, one option is to use SvelteKit itself to do up the docs. (As mentioned on the [SvelteKit docs](https://kit.svelte.dev/docs/packaging), "src/routes might be a documentation or demo site that accompanies the library...".
  - Pros: Code base is simpler with no need for a separate workflow for docs
  - Cons: We will lose all the astro niceties, such as the Code component, navigation (sidebar and ToC) and theming. 
- If sticking with astro, [Astro Starlight](https://starlight.astro.build/), Astro's official documentation framework, may be something to consider. Fingers crossed the existing mdx files should port over nicely
- For a docs site in Astro, do you want to consider changing this to a monorepo setup? Codebases I've been following ([like Svelte](https://github.com/sveltejs/svelte) have been adapting a monorepo setup with pnpm workspaces with the library in a `packages` folder and the docs in a `sites` folder. 
  - Pros: Both the docs and library lives in the same git branch. The current docs import the 'svelte-range-slider-pips' npm package, which means it may get out of sync. Importing from a workspace will mean that we get to see the latest version of the library within the docs without the publish and import steps.
  - Cons: Restricted to using pnpm

**Reply by simeydotme** [reply]
thanks for the great suggestions!

question regarding monorepos, though; would placing multiple packages (library + docs) into a single git repo mean that any time I fixed a bug/typo into the Docs, the version number for the parent package would change (even though, to the consumer, there is no actual change to the library)

I'm not adverse to the idea, because technically I already have this kind of setup running with the `/test/` folder (bad naming, really) 

I think I'd prefer to stick with Astro just because of the nice template and starlight theme. When I built the docs originally the template/theme I used was a fork of starlight (i think) because starlight was still very new and had a lot of limitations and issues... but I believe it's a lot better now and could probably simplify some of the components. 

One thing I was very keen on and put a lot of effort into was having tiny stand-alone components/chunks that both served as the runnable-component, and the code that is displayed as a snippet for copy/paste; https://github.com/simeydotme/svelte-range-slider-pips/blob/astro-pages/src/components/Demos/examples/float/Float.svelte ... this feels really good when building/editing and means that I never accidentally wrote buggy code for copy/paste and I never had to test the code (the test was the demo working)

**Reply by kelvinsjk** [reply]
The package versions will be kept separate. Each package in the monorepo (ie the library itself and the docs site) will have their own `package.json` and hence their own version. 

For my beginner level workflows I kind of use the two folders separately (ie do the `npm publish` inside the library folder itself and then let vercel do their work to handle updates on the docs folder (not sure how github pages will work for that)).

I've seen other authors coordinate the workflow from the root of the monorepo, with its own root package.json, though.

I tried this idea out [in a fork](https://github.com/kelvinsjk/svelte-range-slider-pips), where I

- moved the files and folders in the project root into `packages/svelte-range-slider-pips`
- added a `pnpm-workspace.yaml` file with the following: 
```yaml
packages:
  - 'packages/*'
  - 'sites/*'
```
- moved the files and folders from the root of the branch `astro-pages` into `sites/docs`
- modified the `package.json` file in the docs from
```json
"svelte-range-slider-pips": "^2.2.2",
```
to 
```json
"svelte-range-slider-pips": "workspace:*",
```

Locally running `pnpm dev` on `packages/svelte-range-slider-pips` and on `sites/docs` seem to get the intended results.

**Reply by simeydotme** [reply]
🥺

ooo, i'll have a look at this later :D

**Comment by simeydotme** [comment]
Ok, I just updated the immutable to false (it got lost when I moved to Vite bundler) and have tried to fix the exports.

**Reply by Gurimarukin** [reply]
Default import works again for me. I don't know which one is better, but I'd say default is the normal behaviour for a Svelte component, and it is one less breaking change ^^

Some of my sliders work again, but I have one which still doesn't slide. I'm trying to investigate.

**Reply by simeydotme** [reply]
I think the SvelteKit Library scaffold sets up to have a large number of library components exported by default, so that's probably why it was done that way. I just didn't think to check it 😂
